### DISCUSSION TITLE
Dynamic Mapper ImplementationName

### DISCUSSION THREAD

**Discussion by alantanlc** [mapstruct/mapstruct#3663] [Q&A]
The following generates `Mapper131Impl.java` on maven compilation:
```
@Mapper(componentModel = "spring", implementationName = "<CLASS_NAME>131Impl")
public interface Mapper {
  // ...
}
```

Is there a way to dynamically derive the version instead of hardcoding it in the mapper annotation? Something like this:
```
@Mapper(componentModel = "spring", implementationName = "<CLASS_NAME><VERSION>Impl")
public interface Mapper {
  // ...
}
```

Then `Mapper131FinalImpl.java` is generated when compiled with Mapstruct version 1.3.1.Final defined in pom.xml.

**Comment by filiphr** [comment]
>Is there a way to dynamically derive the version instead of hardcoding it in the mapper annotation?

No there is no such thing that would derive the version of MapStruct when generating the code. 

May I ask what the purpose of that is? Usually you have a single MapStruct version in one module.

**Comment by alantanlc** [comment]
Apologies for the late reply and thanks for the response!

This is no longer required as I've found another way using a Maven multi-module approach, but to elaborate on the initial intent:

I am maintaining mappers as a separate jar module which is used by the main application module.

Was initially thinking that if the Impl class names can be dynamically generated based on the Mapstruct version used, then in the main application module, it can import multiple versions of the same module/artifactId (mapper-module) for feature flag toggling, e.g.

**Main module application pom.xml**
```
<dependencies>

  <!-- Generates MapperImpl131Final.java -->
  <dependency>
    <groupId>com.example</groupId>
    <artifactId>mapper-module</artifactId>
    <version>1.3.1.Final</version>
  <dependency>

  <!-- Generates MapperImpl160.java -->
  <dependency>
    <groupId>com.example</groupId>
    <artifactId>mapper-module</artifactId>
    <version>1.6.0</version>
  <dependency>

</dependencies>
```

**Then in main application:**

```java

MapperImpl131Final m131Final;

MapperImpl160 m160:

public Target mapSourceToTarget(Source source) {

  Target target;

  if (isMapstruct131Final) {
    target = m131Final.map(source);
  } else {
    target = m160.map(source);
  }

  return target;

}
```

**But seems like the same artifactId (mapper-module) cannot be imported more than once using different versions in Maven pom.xml anyway** ([stackoverflow](https://stackoverflow.com/questions/24962607/can-multiple-versions-of-the-same-dependency-be-used-in-a-single-maven-repositor)), so the above approach would not work.

**Instead, using multi-module approach for feature flag toggling instead**:

**Main module application pom.xml**
```
<dependencies>

  <!-- Generates MapperImpl.java -->
  <dependency>
    <groupId>com.example</groupId>
    <artifactId>mapper-module</artifactId>
    <version>1.3.1.Final</version>
  <dependency>

  <!-- Generates Mapper160Impl.java -->
  <dependency>
    <groupId>com.example</groupId>
    <artifactId>mapper-module-160</artifactId>
    <version>1.6.0</version>
  <dependency>

</dependencies>
```

In this approach, **mapper-module-160** module contains Mapper160.java that **extends** from **mapper-module** module's Mapper.java, so that multiple mapstruct versions of mappers can be generated in different modules.
